File: backend/app/api/likes.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session, joinedload
from typing import List
from sqlalchemy import and_, select, desc, or_

from app.core.database import get_db
from app import schemas
from app.models.like import Like
from app.models.track import Track
from app.api.deps import get_current_active_user
from app.models.user import User

router = APIRouter()

@router.get("/", response_model=List[schemas.Like])
def read_user_likes(
    skip: int = 0, 
    limit: int = 100, 
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """
    Получить список лайков текущего пользователя, отсортированный по дате 
    добавления (новые в начале).
    
    - **skip**: смещение для пагинации
    - **limit**: максимальное количество возвращаемых записей
    """
    # Используем joinedload для эффективной загрузки связанных треков
    likes = db.query(Like).options(joinedload(Like.track))\
              .filter(Like.user_id == current_user.id)\
              .order_by(desc(Like.created_at))\
              .offset(skip).limit(limit).all()
    
    return likes


@router.post("/", response_model=schemas.Like, status_code=status.HTTP_201_CREATED)
def create_like(
    like_data: schemas.LikeCreate, 
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """
    Поставить лайк треку.
    
    - **track_id**: ID трека для лайка
    - **artwork_url**: URL обложки трека (опционально)
    """
    # Проверяем существует ли трек
    track = db.query(Track).filter(Track.id == like_data.track_id).first()
    if not track:
        raise HTTPException(status_code=404, detail="Трек не найден")
    
    # Проверяем, не поставил ли уже пользователь лайк этому треку
    existing_like = db.query(Like).filter(
        and_(Like.user_id == current_user.id, Like.track_id == like_data.track_id)
    ).first()
    
    if existing_like:
        raise HTTPException(status_code=400, detail="Вы уже поставили лайк этому треку")
    
    # Получаем URL обложки трека, если он не был передан
    artwork_url = like_data.artwork_url
    if not artwork_url and track.artwork_url:
        artwork_url = track.artwork_url
    
    # Создаем новый лайк
    db_like = Like(
        user_id=current_user.id,
        track_id=like_data.track_id,
        artwork_url=artwork_url
    )
    
    db.add(db_like)
    db.commit()
    db.refresh(db_like)
    
    return db_like


@router.delete("/{track_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_like(
    track_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """
    Удалить лайк трека.
    
    - **track_id**: ID трека, лайк которого нужно удалить
    """
    # Находим лайк
    like = db.query(Like).filter(
        and_(Like.user_id == current_user.id, Like.track_id == track_id)
    ).first()
    
    if not like:
        raise HTTPException(status_code=404, detail="Лайк не найден")
    
    # Удаляем лайк
    db.delete(like)
    db.commit()
    
    return None


@router.get("/check/{track_id}", response_model=bool)
def check_like(
    track_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """
    Проверить, поставил ли пользователь лайк треку.
    
    - **track_id**: ID трека для проверки
    """
    like = db.query(Like).filter(
        and_(Like.user_id == current_user.id, Like.track_id == track_id)
    ).first()
    
    return like is not None


@router.get("/search", response_model=List[schemas.Like])
def search_user_likes(
    query: str = Query(None, min_length=2, description="Поисковый запрос (мин. 2 символа)"),
    skip: int = 0, 
    limit: int = 20, 
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """
    Поиск среди лайкнутых треков пользователя по названию или автору.
    
    - **query**: текст для поиска
    - **skip**: смещение для пагинации
    - **limit**: максимальное количество результатов
    """
    if not query:
        return []
    
    # Используем оператор ILIKE для поиска без учета регистра с частичным совпадением
    search_pattern = f"%{query}%"
    
    try:
        # Поиск среди лайкнутых треков
        likes = db.query(Like).options(joinedload(Like.track))\
                .join(Track, Like.track_id == Track.id)\
                .filter(
                    Like.user_id == current_user.id,
                    or_(
                        Track.title.ilike(search_pattern),
                        Track.artist.ilike(search_pattern)
                    )
                )\
                .order_by(desc(Like.created_at))\
                .offset(skip).limit(limit).all()
        
        return likes
    except Exception as e:
        print(f"Ошибка при поиске лайков: {e}")
        return [] 

File: frontend/src/services/likes.ts
import { ApiService } from './api';
import { API_URL } from '@/config';
import { authService } from './auth';
import { Track } from './tracks';

// Интерфейс для лайка
export interface Like {
  id: string;
  user_id: string;
  track_id: string;
  created_at: string;
  artwork_url?: string | null;
  track?: Track; // Связанный трек
}

// Класс для работы с API лайков
export class LikeService {
  /**
   * Получить все лайки текущего пользователя вместе с информацией о треках
   */
  static async getUserLikes(skip: number = 0, limit: number = 100): Promise<Like[]> {
    try {
      const likes = await ApiService.get<Like[]>(`/likes?skip=${skip}&limit=${limit}`);
      return likes;
    } catch (error) {
      console.error('Ошибка при получении лайков:', error);
      return [];
    }
  }

  /**
   * Поставить лайк треку
   */
  static async likeTrack(trackId: string, artworkUrl?: string): Promise<Like | null> {
    try {
      return await ApiService.post<Like>('/likes', { 
        track_id: trackId,
        artwork_url: artworkUrl
      });
    } catch (error) {
      console.error('Ошибка при добавлении лайка:', error);
      return null;
    }
  }

  /**
   * Удалить лайк с трека
   */
  static async unlikeTrack(trackId: string): Promise<boolean> {
    try {
      // Делаем запрос напрямую через fetch, а не через ApiService
      const accessToken = authService.getAccessToken();
      const response = await fetch(`${API_URL}/likes/${trackId}`, {
        method: 'DELETE',
        headers: {
          ...(accessToken ? { 'Authorization': `Bearer ${accessToken}` } : {})
        }
      });
      
      return response.ok;
    } catch (error) {
      console.error('Ошибка при удалении лайка:', error);
      return false;
    }
  }

  /**
   * Проверить, поставил ли пользователь лайк треку
   */
  static async checkLike(trackId: string): Promise<boolean> {
    try {
      return await ApiService.get<boolean>(`/likes/check/${trackId}`);
    } catch (error) {
      console.error('Ошибка при проверке лайка:', error);
      return false;
    }
  }
} 